### ISSUE TITLE
False positive -- RiskyTruthyFalsyComparison with array shape

### ISSUE THREAD

**Issue by olavmo-sikt** [vimeo/psalm#11481]
When accessing a field in an [array shape](https://psalm.dev/docs/annotating_code/type_syntax/array_types/#array-shapes), we get a `RiskyTruthyFalsyComparison` error indicating that the field can be `null`, even though that is not part of the array shape:

```php
<?php
/**
 * @param array{foo: string} $arg
 * @return void
 */
function foo($arg) {
    // Fails with:
    // RiskyTruthyFalsyComparison - 10:9 - Operand of type null|string contains type string, [...]
    if (empty($arg['foo'])) {}
}
```
(https://psalm.dev/r/88e7bd3f36)

This only happens when checking the field in the array.
If it is assigned to a variable first, no error is produced:

```php
<?php
/**
 * @param array{foo: string} $arg
 * @return void
 */
function foo($arg) {
    $value = $arg['foo'];
    // No error:
    if (empty($value)) {}
}
```
(https://psalm.dev/r/5085a164cc)
